src/Microsoft.AspNetCore.Mvc.Core/Internal/ResourceInvoker.cs

```diff
@@ -663,9 +641,152 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
+                throw new InvalidOperationException(message);
+            }
+
+            await InvokeNextResourceFilter();
+
+            Debug.Assert(_resourceExecutedContext != null);
+            return _resourceExecutedContext;
+        }
+
+        private async Task InvokeNextResourceFilter()
+        {
+            try
+            {
+                var scope = Scope.Resource;
+                var next = State.ResourceNext;
+                var state = (object)null;
+                var isCompleted = false;
+                while (!isCompleted)
+                {
+                    await Next(ref next, ref scope, ref state, ref isCompleted);
+                }
+            }
+            catch (Exception exception)
+            {
+                _resourceExecutedContext = new ResourceExecutedContext(_resourceExecutingContext, _filters)
+                {
+                    ExceptionDispatchInfo = ExceptionDispatchInfo.Capture(exception),
+                };
+            }
+
+            Debug.Assert(_resourceExecutedContext != null);
+        }
+
+        private async Task InvokeNextExceptionFilterAsync()
+        {
+            try
+            {
+                var next = State.ExceptionNext;
+                var state = (object)null;
+                var scope = Scope.Exception;
+                var isCompleted = false;
+                while (!isCompleted)
+                {
+                    await Next(ref next, ref scope, ref state, ref isCompleted);
+                }
+            }
+            catch (Exception exception)
+            {
+                _exceptionContext = new ExceptionContext(_actionContext, _filters)
+                {
+                    ExceptionDispatchInfo = ExceptionDispatchInfo.Capture(exception),
+                };
+            }
+        }
+
+        private async Task InvokeResultFilterPipeline(ResultExecutionScope resultExecutionScope)
+        {
+            var resultNext = State.ResultBegin;
+            var resultScope = Scope.Invoker;
+            var resultState = (object)null;
+            var resultIsCompleted = false;
+
+            while (!resultIsCompleted)
+            {
+                if (resultExecutionScope == ResultExecutionScope.Action)
+                {
+                    await ResultNext<IResultFilter, IAsyncResultFilter>(ref resultNext, ref resultScope, ref resultState, ref resultIsCompleted);
+                }
+                else
+                {
+                    Debug.Assert(
+                        resultExecutionScope == ResultExecutionScope.Authorization ||
+                        resultExecutionScope == ResultExecutionScope.Exception ||
+                        resultExecutionScope == ResultExecutionScope.Resource);
+                    await ResultNext<IAlwaysRunResultFilter, IAsyncAlwaysRunResultFilter>(ref resultNext, ref resultScope, ref resultState, ref resultIsCompleted);
+                }
+            }
```

Comment:
This doesn't seem super right. I'd rather see this broken up into two separate methods that call `ResultNext<>`. I don't like the `ResultExecutionScope` concept since it's just used for this `if`